### PR TITLE
Fix container volume permission issues with docker overlayfs

### PR DESCRIPTION
**Pull request by joxl**
## Summary

Running tests locally with docker was crashing due to insufficient file permissions of the HQ directory structure.  The solution is to recursively chmod the entire overlayfs mount, allowing world-read (and execute, where required) access.  This allows the cchq user to access all files within, regardless of how restricted they are on the source-OS (dependent on OS and user-specific umask setting and/or manual chmods).

To accomplish this, `docker/run.sh` observes two new environment variables when configuring an `overlayfs` overlay mount:

- `DOCKER_HQ_OVERLAYFS_CHMOD=[yes|no]`: Perform a recursive chmod on HQ dir to ensure read access for `cchq` user.
- `DOCKER_HQ_OVERLAYFS_METACOPY=[on|off]`: Set the `metacopy=on` mount option for the `overlayfs` mount. This is a performance optimization for the chmod operation.

Since docker tests have been working (for travis, at least) without this modification, the changes are disabled when absent and only activate when explicitly configured.

Easiest to review commit by commit.

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### QA Plan

No QA.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

**Comment by joxl** [comment]
Looks like this PR didn't justify a travis build. Not sure how that works. Is there a way to force one?

**Comment by joxl** [comment]
> Looks like this PR didn't justify a travis build.

Nevermind, there is one now. Strange.

**Review by snopoke — COMMENTED** [review]
I think only @millerdev has enough context on this to really review. 

One comment is that it would be good to have a comment about WHY we have to support these three states.

**Comment by joxl** [comment]
> would be good to have a comment about WHY we have to support these three states.

I'll add these comments.

My suspicion is that is that it `aufs` is not needed anywhere, but I didn't want to remove it in this PR to keep this a low-risk "fix" change set.

Looking at  #12154, `overlayfs` was a later addition; `aufs` being the default (perhaps only) Docker overlay engine when the script was originally written.  I'm speculating here, but I suspect `aufs` was not ripped out back then because `overlayfs` would not have been very mature, and perhaps not even widely supported.

File: docker/run.sh [hunk 1]
```diff
@@ -231,20 +267,84 @@ if [ "$DOCKER_HQ_OVERLAY" == "none" ]; then
...
+            now=$(date +%s)
+            # add world-read (and world-x for dirs and existing-x files)
+            chmod -R o+rX commcare-hq
+            delta=$(($(date +%s) - $now))
+            echo "(delta=${delta}sec)" >&2  # append the previous log line
+        fi
     else
-        mount -t aufs -o br=lib/overlay:commcare-hq-ro none commcare-hq
+        # TODO: use overlayfs and drop support for aufs
+        mount -t aufs -o br=/mnt/lib/overlay:/mnt/commcare-hq-ro none commcare-hq
```
Comment (millerdev) [inline docker/run.sh:308]: I don't think anyone is using this in a dev setup, so if this is not required for Travis it's probably fine to drop it.

**Review by millerdev — APPROVED** [review]

File: docker/run.sh [hunk 2]
```diff
@@ -7,8 +7,51 @@ if [ -z "$1" ]; then
     exit 0
 fi
 
+function logmsg {
```
[the remark this answers is not in the document]
Comment (joxl) [reply docker/run.sh]: Sure. Actually, I moved them to a separate file which is now sourced. This makes them available in other places as well.

File: docker/run.sh (hunk 1 above)
Comment (joxl) [reply docker/run.sh:237]: I'll test this and put it in a separate PR.

**Review by millerdev — APPROVED** [review]

**Pull request merged**